common/authorization/interceptor.go

```diff
@@ -68,183 +68,217 @@ var (
-			}
-		}
+// TLSInfoFromContext extracts TLS information from the context's peer value.
+func TLSInfoFromContext(ctx context.Context) *credentials.TLSInfo {
+	p, ok := peer.FromContext(ctx)
+	if !ok {
+		return nil
 	}
+	if tlsInfo, ok := p.AuthInfo.(credentials.TLSInfo); ok {
+		return &tlsInfo
+	}
+	return nil
+}
 
-	if a.authorizer != nil {
-		var namespace string
-		requestWithNamespace, ok := req.(hasNamespace)
-		if ok {
-			namespace = requestWithNamespace.GetNamespace()
-		}
-
-		handler := a.getMetricsHandler(metrics.AuthorizationScope, namespace)
-		result, err := a.authorize(ctx, claims, &CallTarget{
-			Namespace: namespace,
-			APIName:   info.FullMethod,
-			Request:   req,
-		}, handler)
-		if err != nil {
-			handler.Counter(metrics.ServiceErrAuthorizeFailedCounter.GetMetricName()).Record(1)
-			a.logAuthError(err)
-			return nil, errUnauthorized // return a generic error to the caller without disclosing details
-		}
-		if result.Decision != DecisionAllow {
-			handler.Counter(metrics.ServiceErrUnauthorizedCounter.GetMetricName()).Record(1)
-			// if a reason is included in the result, include it in the error message
-			if result.Reason != "" {
-				return nil, serviceerror.NewPermissionDenied(RequestUnauthorized, result.Reason)
-			}
-			return nil, errUnauthorized // return a generic error to the caller without disclosing details
-		}
+// PeerCert extracts an x509 certificate from given tlsInfo.
+func PeerCert(tlsInfo *credentials.TLSInfo) *x509.Certificate {
+	if tlsInfo == nil || len(tlsInfo.State.VerifiedChains) == 0 || len(tlsInfo.State.VerifiedChains[0]) == 0 {
+		return nil
 	}
-	return handler(ctx, req)
+	// The assumption here is that we only expect a single verified chain of certs (first[0]).
+	// It's unclear how we should handle a situation when more than one chain is presented,
+	// which subject to use. It's okay for us to limit ourselves to one chain.
+	// We can always extend this logic later.
+	// We take the first element in the chain ([0]) because that's the client cert
+	// (at the beginning of the chain), not intermediary CAs or the root CA (at the end of the chain).
+	return tlsInfo.State.VerifiedChains[0][0]
 }
 
-func (a *interceptor) authorize(
-	ctx context.Context,
-	claims *Claims,
-	callTarget *CallTarget,
-	metricsHandler metrics.Handler) (Result, error) {
-	startTime := time.Now().UTC()
-	defer func() {
-		metricsHandler.Timer(metrics.ServiceAuthorizationLatency.GetMetricName()).Record(time.Since(startTime))
-	}()
-	return a.authorizer.Authorize(ctx, claims, callTarget)
+// HeaderGetter is an interface for getting a single header value from a case insensitive key.
+type HeaderGetter interface {
+	Get(string) string
 }
 
-func (a *interceptor) logAuthError(err error) {
-	a.logger.Error("Authorization error", tag.Error(err))
+// Wrapper for gRPC metadata that exposes a helper to extract a single metadata value.
+type grpcHeaderGetter struct {
```

Comment:
This looks like it might be redundant with the helper we have here https://github.com/temporalio/temporal/blob/main/common/headers/headers.go#L60

Is there a reason we need this separate wrapper?

Reply:
Ah I see because the http request Header object has a different format. Still, I think it would make sense to keep all the header utilities together. Not a big deal though, this is fine too.